feature-login

shahn95 opened the pull request:
Story:
[As a user I should be able to login with the correct credentials#8](https://github.com/nilenso/chaat/issues/8)

This PR implement authentication and authorization using JWT's. This is currently only useful when sending requests to the `delete-user` handler. You need to be authenticated as `username` to have authorization to delete `username`.

Note to reviewers:  If you go through the commits, do not go through all 108 commits. Please start reviewing from commit id `6266904` `Wrap :delete users endpoint with authentication middleware`. I did some things in git that is possibly causing github to incorrectly compute the commit diff. The code diff is correct though.

Chaat so far: We have user account creation/deletion, and login (this PR). I'm currently working on friend requests. There is no functionality to send messages yet. Everything runs locally.

shahn95 committed:
Test commit

shahn95 committed:
Update project.clj dependencies

shahn95 committed:
Initialize docker-compose file for local postgresql db

shahn95 committed:
Update Dockerfile

shahn95 committed:
Initialize config.clj to parse config.edn

shahn95 committed:
Initialize config file

shahn95 committed:
Initialize up/down migration files for users table

shahn95 committed:
Initialize migrations.clj to apply/rollback migrations

shahn95 committed:
Initialize model/user.clj with wip user account create/delete functions

shahn95 committed:
Add wrap-params, wrap-keyword-params, and wrap-json-params middleware

shahn95 committed:
Add routes for signup and delete-user, and filter on request methods

shahn95 committed:
Update handler.clj with handlers for signup and delete-account

shahn95 committed:
Update intro

foo

shahn95 committed:
Update health-check-test

shahn95 committed:
Update dependencies

shahn95 committed:
Start using a map return structure and move username/password validation from model -> handler layer

This is wip, and a checkpoint to document progress. I plan to refactor this later to try a tuple return
structure, and have different validations at both handler and model layers. More on that in a future commit.

Started implementing a 3 key map (:success, :message, :description) to unify return values from validation
functions and db operations. Goal is to simplify passing success/error info back to the handler layer.

Created fn build-result-map to simplify construction of the returned map.
Created fn send-response to send a good/bad response based on the returned map.
Created fn gen-new-user-map to generate a map with all data required to insert a new user into the db.

shahn95 committed:
Create handler folder and refactor with tuple return structure. Move validation into validation.clj

Moved validation out of handler.clj and into validation.clj
Moved handler.clj and validation.clj into handler folder.

Refactored handler.clj to use a tuple return structure for error handling and validation:
[params nil] or [nil err-msg]

Updated handler level validation to only check that required parameters exist.
I plan to add more robust validation using a library like schema in a future commit.

shahn95 committed:
Refactor with tuple return structure and remove db queries. Add model layer validation.

Removed db queries from this layer will go into a separate db layer in the next commit.

shahn95 committed:
Initialize db layer with user related (new?/add/delete) db queries

Creates a connection pooled datasource and performs queries on the user table to
check if users exists along with adding/deleting users.

This also uses a tuple return structure.

shahn95 committed:
Update to use callable functions: run-migrations and rollback

shahn95 committed:
Update to use callable functions to fetch config info

shahn95 committed:
Add config to setup a different db instance to run tests on

shahn95 committed:
Update dependencies

shahn95 committed:
Add start/stop/restart functions, update middleware order, and use local server port from config.

shahn95 committed:
Update new-user? function to receive datasource as a parameter

This is necessary if new-user? is to be used within a db transaction

shahn95 committed:
Refactor error handling/validation to use a map return structure

Was previously using a tuple structure. Now changed to use a
{:result result :error error} structure.

Made a errors namespace to house the do-or-error function.

shahn95 committed:
Update password encryption to remove manual salting, bcrypt already does this

Updated gen-new-user-map in user.clj to remove manual salting procedure
Removed static salt from config.edn
Removed get-static-salt fn from config.clj
Updated 001-users.up.sql to remove dynamic_salt column from users table

shahn95 committed:
Fix postgres timestamp error by requiring next.jdbc.date-time

shahn95 committed:
Refactor to use component libary: add server component

Started incremental refactor to use the component library.
Implemented a component system map with a server component.
Will add more components in future commits.

shahn95 committed:
Refactor to use component library: add db component

Added a db component to the component system map. db component is injected into the
make-app function, and eventually used for db queries.

shahn95 committed:
Add ring/ring-mock dependency

shahn95 committed:
Add local test port number to config file

shahn95 committed:
Remove ragtime.core dependency and edit comments

shahn95 committed:
Add fn to get local test port from config file

shahn95 committed:
Improve docstring for fn do-or-error and add comments

shahn95 committed:
Change run-migrations and rollback fns to take dbspec as an argument

Will improve these further in a future commit.

shahn95 committed:
Add comments

shahn95 committed:
Add a user-exists? fn

shahn95 committed:
Improve comment to explain handler level validation

shahn95 committed:
Add docstrings

shahn95 committed:
Change jt/local-time -> jt/instant for health check

shahn95 committed:
Modify delete-user successful response message

shahn95 committed:
Modify namespace aliases

shahn95 committed:
Modify namespace alias

shahn95 committed:
Edit comments and explain model layer validation

shahn95 committed:
Initialize fixture namespace with test-system and test-fixture for running tests

shahn95 committed:
Add unit tests with stub fns to test handler layer functions

Moved into handler folder to mimic project structure.
Will add integration tests in a future commit

shahn95 committed:
Add units tests with stub fns to test model layer user functions

Will add integration tests in a future commit

shahn95 committed:
Add integration tests for db layer user functions

Will add tests for query results in a future commmit

shahn95 committed:
Update readme to remove misleading info

Will add more info in a future commit.

shahn95 committed:
Add dependencies

shahn95 committed:
Initialize app_test.clj with end-end health check test

shahn95 committed:
Add comment about improving fixtures

shahn95 committed:
Improve parameter names for new-user? and user-exists? functions

shahn95 committed:
Add helper function get-time-instant which calls jt/instant

shahn95 committed:
Improve db layer tests

shahn95 committed:
Improve model.user layer tests

Added integration tests and removed stubs where not necessary

shahn95 committed:
Improve handler layer tests

Add integration tests and remove stubs where not necessary

shahn95 committed:
Initialize db/utils.clj with database health-check

shahn95 committed:
Modify health-check endpoint to also test for db status

This makes the health-check endpoint a more comprehensive app status check,
and is nice to use for an end->end test.

Modified routes.clj, handler.clj and handler_test.clj for this change.

shahn95 committed:
Update dependencies

shahn95 committed:
Wrap :delete users endpoint with authentication middleware

Authorization will be required to delete a user

shahn95 committed:
Add secret for signing JWT's to config file

shahn95 committed:
Add until-err-> macro to make do-or-error flow in code cleaner

Work-in-progress macro

shahn95 committed:
Update fn name from validate-signup-details -> validate-credentials

shahn95 committed:
Add login endpoint and send-reponse helper function

shahn95 committed:
Update model layer user validation functions

shahn95 committed:
Add model layer user login and authenticate functions

shahn95 committed:
Add get-password-hash function to retrieve username's pwd hash from db

shahn95 committed:
Use updated name for validation function call in signup handler

shahn95 committed:
Update delete-user handler to check for user authorization before deleting a user

Also added helper function is-auth-user

shahn95 committed:
Add tests for get-password-hash in db layer

shahn95 committed:
Add tests for authenticate and login fns in model layer

shahn95 committed:
Add test for login handler

shahn95 committed:
Update delete-user handler test to include authorization tests

shahn95 committed:
Update comments and remove commented code

shahn95 committed:
Update testing context strings

shahn95 committed:
Correct typo in  testing context sting

shahn95 committed:
Remove unnecessary comments

qptr commented on docker-compose.yml:
```diff
@@ -0,0 +1,24 @@
+      TZ: Etc/UTC
+      PGTZ: Etc/UTC
+      POSTGRES_PASSWORD: password
+      POSTGRES_USER: chaat_dev
```
Should we be putting secrets elsewhere?

qptr commented on src/chaat/app.clj:
```diff
@@ -1,30 +1,73 @@
-  (.stop @server))
+(defn start []
+  ;; can add logging and other setup
+  ;; (component/start chaat-system)
```
stray comment

shahn95 committed:
Rename fn get-password-hash to get-user-details and fetch all user details

shahn95 committed:
Update jwt payload to include correct issue/expiry times

shahn95 committed:
Update login and authenticate fns to use db.user/get-user-details

JWT payload will now also include the correct user uuid from the db.

shahn95 committed:
Remove commented code

shahn95 committed:
Update .gitignore

shahn95 committed:
Remove and cleanup comments

shahn95 committed:
Improve get config fns

shahn95 committed:
Update implementation of user-exists? and new-user? fns

shahn95 committed:
Update to return correct status codes for errors

Created an error map in the chaat.handler.errors namespace which
contains status codes for common errors.

Created a send-response helper fn.

Also updated tests.

shahn95 committed:
Simplify fn names: gen-new-user-map -> gen-new-user and gen-config-map -> gen-config

shahn95 committed:
Removing empty deftest blocks for fns which do not need testing

shahn95 committed:
Remove dead code relating to user display-picture implementation

shahn95 committed:
Update fn name end-end-test -> health-check-test

shahn95 committed:
Update status codes

shahn95 committed:
Update send-response helper function

shahn95 committed:
Fix get-secret fn to work with new get-config fn

shahn95 committed:
Add unauthorized-action to error-table

shahn95 committed:
Make fn authenticate private

shahn95 committed:
Modify to return error from error-table

shahn95 committed:
Add repl commands to comment

shahn95 committed:
Use get-time-instant instead of jt/instant

shahn95 committed:
Update model.user tests: Remove authenticate-test and fix login-test

Also updated time-stub fn

shahn95 committed:
Formatting fix

shahn95 committed:
Update until-err-> macro

shahn95 committed:
Updates to formatting, comments and comment blocks

shahn95 committed:
Fix fn call: gen-new-user-map -> gen-new-user

shahn95 committed:
Update login and delete-user tests, test for correct status codes

shahn95 committed:
Remove get-password-hash-test and add get-user-details test

shahn95 committed:
Merge branch 'main' into feature-login

vkrmis commented on resources/migrations/002-friend-requests.up.sql:
```diff
@@ -0,0 +1,9 @@
+CREATE TABLE users (
+    id                      uuid DEFAULT gen_random_uuid(),
+    from                    uuid,
+    to                      uuid,
+    status                  text,
+    message                 text,
+    sent_timestamp          timestamptz,
+    PRIMARY KEY (id)
+);
```
Is this meant to be here? Looks like the first migration already sets up this table.

shahn95 replied:
This file was not supposed to be here, I will be more careful.

I was trying to fix some issues with github showing old commits from way before this branch in this PR. The issue was caused by branching off from branches which were not yet merged into main. This led to some experimentation, and these files were merged in during that process.

I have deleted these files now.

vkrmis commented on resources/migrations/002-friend-requests.down.sql:
Add the down migration as well? If this migration isn't meant to be reversible, it should throw an error indicating that.

shahn95 replied:
Answered above. I have deleted these files now.

vkrmis commented on src/chaat/handler/handler.clj:
```diff
@@ -30,17 +40,25 @@
   (let [params (:params request)
         {:keys [username password]} params
-        result (handler.validation/validate-signup-details username password)
+        result (handler.validation/validate-credentials username password)
         result (do-or-error result model.user/create db username password)]
```
Can you not use the `until-err->` macro for this? Shadowing bindings like this is a smell imo, shouldn't be needed when you can use a threading macro instead

shahn95 replied:
We discussed my reasons for doing this at the office, and you suggested some good alternatives. I won't do this refactor now, but I am aware of the alternative options including the pedestal library. 

Next time we speak, I would like to spend a few more minutes discussing the details of the other option where we do not return detailed error messages for the failure case, and just return `nil` or let exceptions be thrown which ring will handle. 

I also want to checkout [failjure](https://github.com/adambard/failjure).

vkrmis commented on test/chaat/handler/handler_test.clj:
```diff
@@ -58,18 +82,27 @@
+                request {:params params
+                         :identity identity}
+                response (handler/delete-user datasource request)]
+            (is (= 404 (:status response)))))))))
```
Could use another test to verify that expired tokens (7+ days old) fail authorization checks

shahn95 replied:
This would be a good test to add. I am parking chaat now, so will not make this change, but will keep this in mind for future projects.

vkrmis commented on src/chaat/routes.clj:
```diff
@@ -1,10 +1,17 @@
         ["health-check" {:get #(handler/health-check db %)}]
         ["users" {:post #(handler/signup db %)
-                  :delete #(handler/delete-user db %)}]
+                  :delete (wrap-authentication #(handler/delete-user db %) backend)}]
```
Should this be a `wrap-authorization` instead?
Worth looking into how the two are different, since they're frequently used synonymously in conversation (incorrectly)

shahn95 replied:
I'm using [buddy-auth](https://funcool.github.io/buddy-auth/latest/).

The way this currently works and my authentication/authorisation understanding is as follows:
- obtain a JWT from the `login` handler by passing valid user credentials. I think of this as user authentication. 
- `wrap-authentication` middleware will add an `:identity` key to requests with an authorisation header containing a valid JWT. I think of this as authenticating the request. 
- `delete-user` handler will check if the request's authenticated identity is the same as the user being deleted. This is where the authorisation happens.

I could also do this in a slightly different way by wrapping the `delete-user` handler in both `wrap-authorization` and `wrap-authentication`. Wrapping the handler in `wrap-authorization` allows me to use the `throw-unauthorized` function in the handler which the middleware will catch and handle in a default or custom manner.

shahn95 committed:
Removing friend request up/down migrations, unrelated to this PR

These were test files which unintentionally made their way into this PR while
we were doing some git experimentation. Fixing this by removing the files.

siripr4 commented on src/chaat/config.clj:
```diff
@@ -29,6 +29,9 @@
   (:local-test-port (:app (get-config))))
 
+(defn get-secret []
+  (:secret (:app (get-config))))
```
```suggestion
  (get-in get-config [:app :secret]))
```

shahn95 replied:
Incorporated, thanks!

siripr4 commented on src/chaat/db/user.clj:
```diff
@@ -33,6 +33,18 @@
+  "Retrieve user details for given username from user table"
+  [db username]
+  (jdbc/with-transaction [tx (db)]
+    (if (user-exists? tx username)
```
question: was an alternative approach considered for this, where you don't do the exists check, and derive that from the query result?

shahn95 replied:
Yes, the alternative approach was to just rely on the query result. The same situation arises when we insert a new user into the db. We have a `UNIQUE` constraint on `username` but we still decided to add a `new-user?` check.

I think these checks are good, but they should be in the model layer instead. If I was to continue work on chaat, I would move all these checks to the model layer. Then, the db layer would look closer to what you expected, and we could perform the query directly.

siripr4 commented on src/chaat/errors.clj:
```diff
@@ -1,16 +1,40 @@
+;; :result key value from the result map of the previous expression.
+
+(defn do-result-or-error
+  " Same as do-or-error but allows for the placement of an optional :result key
```
nit-pick (non-blocking): Is the space at the beginning of the docstring intended? (Same on line 4 as well)

shahn95 replied:
Unintended. The auto-format on save wants to put a space at the beginning and end of multi-line docstrings :O

Good catch, I don't like these spaces.

siripr4 commented on test/chaat/handler/handler_test.clj:
```diff
@@ -44,6 +50,24 @@
             response (handler/signup datasource request)]
         (is (= 409 (:status response)))))))
 
+(deftest login-test
```
question: I notice that all the api tests only check for the response status code and not the response itself. Was this a deliberate choice? 

Also, I'd recommend adding tests for incorrect password and bad requests to `login-test`

shahn95 replied:
We were still having discussions on what the ideal response content should be. Hence, I made the deliberate choice to only check status codes at this point in time. 

Good catch, I added an incorrect password test to `login-test`.

siripr4 commented on src/chaat/errors.clj:
```diff
@@ -1,16 +1,40 @@
+to pass the :result key value from the result map of the previous expression "
+  [{:keys [result error]} f & args]
+  (if-not error
+    (apply f (map #(if (= % :result) result %) args))
```
question: What is the anonymous fn `#(if (= % :result) result %)` intended to do? 

Also consider extracting this to a private fn with a name that makes this more explicit

shahn95 replied:
`(map #(if (= % :result) result %) args)` iterates over items in `args` and replaces `:result` with `result`. In this scope, the `:result` key value from the result map was deconstructed into `result`.
This allows us to place a `:result` key in the args list to an expression which essentially passes in the `:result` key value from the result map of the previous threaded expression. 

This is written in a confusing way, and the naming makes it more confusing. It was mainly an experiment to see if we could improve my clunky `do-or-error` flow. If chaat was to continue, this flow would likely be replaced by a library or simplified.

I'll be happy to discuss why this felt necessary at that time when we next speak.

siripr4 reviewed (commented):
nit-pick: There are some unused requires in multiple namespaces, would recommend cleaning those up.

shahn95 committed:
Use get-in to get keys from nested maps

shahn95 committed:
Add test for incorrect password to login-test

shahn95 committed:
Update error status codes

shahn95 commented:
Parking chaat development to shift focus to CH. I have addressed most of the discussion and made some changes. Will be merging this branch.
